DllImportGenerator/Microsoft.Interop.SourceGeneration/Marshalling/MarshallingGeneratorFactory.cs

```diff
@@ -9,133 +7,58 @@
-        /// <param name="info">Object to marshal</param>
-        /// <param name="context">Code generation context</param>
-        /// <returns>Argument syntax for <paramref name="info"/></returns>
-        ArgumentSyntax AsArgument(TypePositionInfo info, StubCodeContext context);
-
-        /// <summary>
-        /// Generate code for marshalling
-        /// </summary>
-        /// <param name="info">Object to marshal</param>
-        /// <param name="context">Code generation context</param>
-        /// <returns>List of statements to be added to the P/Invoke stub</returns>
-        /// <remarks>
-        /// The generator should return the appropriate statements based on the
-        /// <see cref="StubCodeContext.CurrentStage" /> of <paramref name="context"/>.
-        /// For <see cref="StubCodeContext.Stage.Pin"/>, any statements not of type
-        /// <see cref="FixedStatementSyntax"/> will be ignored.
-        /// </remarks>
-        IEnumerable<StatementSyntax> Generate(TypePositionInfo info, StubCodeContext context);
-
-        /// <summary>
-        /// Returns whether or not this marshaller uses an identifier for the native value in addition
-        /// to an identifer for the managed value.
-        /// </summary>
-        /// <param name="info">Object to marshal</param>
-        /// <param name="context">Code generation context</param>
-        /// <returns>If the marshaller uses an identifier for the native value, true; otherwise, false.</returns>
-        /// <remarks>
-        /// <see cref="StubCodeContext.CurrentStage" /> of <paramref name="context"/> may not be valid.
-        /// </remarks>
-        bool UsesNativeIdentifier(TypePositionInfo info, StubCodeContext context);
-
-        /// <summary>
-        /// Returns if the given ByValueContentsMarshalKind is supported in the current marshalling context.
-        /// A supported marshal kind has a different behavior than the default behavior.
+        /// Create an <see cref="IMarshallingGenerator"/> instance for marshalling the supplied type in the given position.
         /// </summary>
-        /// <param name="marshalKind">The marshal kind.</param>
-        /// <param name="context">The marshalling context.</param>
-        /// <returns></returns>
-        bool SupportsByValueMarshalKind(ByValueContentsMarshalKind marshalKind, StubCodeContext context);
+        /// <param name="info">Type details</param>
+        /// <param name="context">Metadata about the stub the type is associated with</param>
+        /// <returns>A <see cref="IMarshallingGenerator"/> instance.</returns>
+        public IMarshallingGenerator Create(
+            TypePositionInfo info,
+            StubCodeContext context);
     }
 
-    /// <summary>
-    /// Exception used to indicate marshalling isn't supported.
-    /// </summary>
-    internal class MarshallingNotSupportedException : Exception
+    public class DefaultMarshallingGeneratorFactory<TOptions> : IMarshallingGeneratorFactory
```

Comment:
Create these inheritance chains makes things difficult for changes. Can we try another pattern?

Reply:
Would a model more similar to the decorator pattern (composition) be preferable?